task/microsoft/task.go

```diff
@@ -0,0 +1,329 @@
+		return err
+	}
+	log.Println("[INFO] Reading VirtualNetwork...")
+	if err := t.Resources.VirtualNetwork.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading SecurityGroup...")
+	if err := t.Resources.SecurityGroup.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading Subnet...")
+	if err := t.Resources.Subnet.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading VirtualMachineScaleSet...")
+	if err := t.Resources.VirtualMachineScaleSet.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Done!")
+	t.Attributes.Addresses = t.Resources.VirtualMachineScaleSet.Attributes.Addresses
+	t.Attributes.Status = t.Resources.VirtualMachineScaleSet.Attributes.Status
+	t.Attributes.Events = t.Resources.VirtualMachineScaleSet.Attributes.Events
+	return nil
+}
+
+func (t *Task) Delete(ctx context.Context) error {
+	log.Println("[INFO] Downloading Directory...")
+	if t.Attributes.Environment.Directory != "" && t.Read(ctx) == nil {
+		if err := t.Pull(ctx, t.Attributes.Environment.Directory); err != nil && err != universal.NotFoundError {
+			return err
+		}
+	}
+	log.Println("[INFO] Deleting VirtualMachineScaleSet...")
+	if err := t.Resources.VirtualMachineScaleSet.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting Subnet...")
+	if err := t.Resources.Subnet.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting SecurityGroup...")
+	if err := t.Resources.SecurityGroup.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting VirtualNetwork...")
+	if err := t.Resources.VirtualNetwork.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting BlobContainer...")
+	if err := t.Resources.BlobContainer.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting StorageAccount...")
+	if err := t.Resources.StorageAccount.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting ResourceGroup...")
+	if err := t.Resources.ResourceGroup.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Done!")
+	return nil
+}
+
+func (t *Task) Logs(ctx context.Context) ([]string, error) {
+	if err := t.Read(ctx); err != nil {
+		return nil, err
+	}
+
+	remote, err := fs.NewFs(ctx, (*t.DataSources.Credentials.Resource)["RCLONE_REMOTE"])
+	if err != nil {
+		return nil, err
+	}
+
+	entries, err := remote.List(ctx, "/log/task")
+	if err != nil {
+		return nil, err
+	}
+
+	var logs []string
+	for _, entry := range entries {
+		object, err := remote.NewObject(ctx, entry.Remote())
+		if err != nil {
+			return nil, err
+		}
+		reader, err := object.Open(ctx)
+		if err != nil {
+			return nil, err
+		}
+		buffer := new(bytes.Buffer)
+		if _, err := io.Copy(buffer, reader); err != nil {
+			return nil, err
+		}
+		logs = append(logs, buffer.String())
+		reader.Close()
+	}
+
+	return logs, nil
+}
+
+func (t *Task) Pull(ctx context.Context, destination string) error {
```

Comment:
this function is pretty close to the one below

Reply:
Yes. Ideally, those functions should be mixed in from a separate package with code that is common to every machine provider. Looks like a nice follow–up.